### PR TITLE
Bazel 0.23 announcement

### PR DESCRIPTION
**Pull request by laurentlb**

### CODE REVIEW

File: _posts/2019-02-26-bazel-0.23.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+title: "Bazel 0.23"
+authors:
+  - laurentlb
+---
+
+[Bazel 0.23.0](https://github.com/bazelbuild/bazel/releases/tag/0.23.0) has
+arrived with an exciting mix of new features, important bug fixes, and updated
+documentation.
+
+This version of Bazel has been regression-tested against 51 open-source
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]: is there a link to a full list of projects?

File: _posts/2019-02-26-bazel-0.23.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+projects, including Gerrit, Protocol Buffers, TensorFlow and the full suite of
+rules hosted in bazelbuild's GitHub organization. This makes it our most
+comprehensively tested release yet. Want to get your open-source project on
+Bazel's CI to ensure that we catch any regressions or have suggestions what we
+should test against? Reach out to
+[bazel-dev@googlegroups.com](mailto:bazel-discuss@googlegroups.com) and we can
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]: looks like there's already a template in bazelbuild/continuous-integration:

https://github.com/bazelbuild/continuous-integration/issues/new/choose

File: _posts/2019-02-26-bazel-0.23.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+*   Introduced the
+    [`--host_compiler`](https://docs.bazel.build/versions/master/command-line-reference.html#flag--host_compiler)
+    flag to allow setting a compiler for host compilation when
+    `--host_crosstool_top` is specified.
+*   Toolchain resolution for cc rules has been enabled via the incompatible flag
+    [`--incompatible_enable_cc_toolchain_resolution`](https://github.com/bazelbuild/bazel/issues/7260).
+    The legacy flag, --enabled_toolchain_types, has been deprecated and will be
+    removed.
+*   [`cc_common.create_cc_toolchain_config_info`](https://docs.bazel.build/versions/master/skylark/lib/cc_common.html#create_cc_toolchain_config_info)
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
*   [`cc_common.create_cc_toolchain_config_info`](https://docs.bazel.build/versions/0.23.0/skylark/lib/cc_common.html#create_cc_toolchain_config_info)
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+*   Added `--host_swiftcopt` for custom flags to pass to swiftc when building tools written in Swift in the host configuration.
+*   Changed the default `apple_platform_type` value from iOS to macOS. You no
+    longer need to pass `--apple_platform_type=macos` when building targets on
+    macOS that previously required it.
+
+
+## C++
+
+*   Introduced the
+    [`--host_compiler`](https://docs.bazel.build/versions/master/command-line-reference.html#flag--host_compiler)
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
    [`--host_compiler`](https://docs.bazel.build/versions/0.23.0/command-line-reference.html#flag--host_compiler)
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    providers instead. See the documentation about [Migrating from Legacy
+    Providers](https://docs.bazel.build/versions/master/skylark/rules.html#migrating-from-legacy-providers).
+*   Some old flags have been removed: `--incompatible_range_type`,
+    `--incompatible_disallow_slash_operator`,
+    `--incompatible_disallow_conflicting_providers`.
+
+
+## Android
+
+*   For [`android_local_test`](https://docs.bazel.build/versions/master/be/android.html#android_local_test),
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
*   For [`android_local_test`](https://docs.bazel.build/versions/0.23.0/be/android.html#android_local_test),
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    [`--host_compiler`](https://docs.bazel.build/versions/master/command-line-reference.html#flag--host_compiler)
+    flag to allow setting a compiler for host compilation when
+    `--host_crosstool_top` is specified.
+*   Toolchain resolution for cc rules has been enabled via the incompatible flag
+    [`--incompatible_enable_cc_toolchain_resolution`](https://github.com/bazelbuild/bazel/issues/7260).
+    The legacy flag, --enabled_toolchain_types, has been deprecated and will be
+    removed.
+*   [`cc_common.create_cc_toolchain_config_info`](https://docs.bazel.build/versions/master/skylark/lib/cc_common.html#create_cc_toolchain_config_info)
+    is now stable and available for production use. See updated [`cc_toolchain
+    tutorial`](https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html)
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
    tutorial`](https://docs.bazel.build/versions/0.23.0/tutorial/cc-toolchain-config.html)
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    and the [reference](https://docs.bazel.build/versions/master/cc-toolchain-config-reference.html).
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
    and the [reference](https://docs.bazel.build/versions/0.23.0/cc-toolchain-config-reference.html).
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+    flag has been added. This flag forces Bazel to use JDK 11 as its default
+    `--host_javabase`.
+*   The [`--incompatible_generate_javacommon_source_jar`](https://github.com/bazelbuild/bazel/issues/5824)
+    flag has been added. This flag forces `java_common.compile` to always generate
+    a source .jar file.
+*   The [`--incompatible_use_toolchain_providers_in_java_common`](https://github.com/bazelbuild/bazel/issues/7186)
+    flag has been added. This flag forces Bazel to pass JavaToolchainInfo and
+    JavaRuntimeInfo providers to java_common APIs instead of configured targets.
+*   An optional `output_source_jar` parameter has been added to
+    [`java_common.compile`](https://docs.bazel.build/versions/master/skylark/lib/java_common.html#compile).
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
    [`java_common.compile`](https://docs.bazel.build/versions/0.23.0/skylark/lib/java_common.html#compile).
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+*   The `--python_version` flag will soon no longer override the declared
+    version of a `py_binary` or `py_test` target. This allows Python 3 binaries
+    to have data dependencies on Python 2 binaries and vice versa. You can
+    preview this new behavior with
+    [`--incompatible_allow_python_version_transitions`](https://github.com/bazelbuild/bazel/issues/7307).
+
+
+## aquery
+
+*   Documentation for the `aquery` command is now available at  https://docs.bazel.build/versions/0.23.0/aquery.html.
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
*   Documentation for the `aquery` command is [now available](https://docs.bazel.build/versions/0.23.0/aquery.html).
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    function type (inputs, outputs, mnemonic) in `aquery` is now allowed.
+*   The content of parameter files can now be displayed in `aquery` output using
+    the  `--include_param_files` flag.
+*   `aquery` text output now highlights if an artifact is a
+    [TreeArtifact](https://github.com/bazelbuild/bazel/issues/7225).
+
+
+## Caching & Execution
+
+*   Locally-executed spawns tagged "no-cache" no longer upload their outputs to
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
*   Locally-executed spawns tagged `no-cache` no longer upload their outputs to
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    doesn't have write access to the cache.
+*   Fixed a longstanding bug in the HTTP remote cache where the value passed to
+    `--remote_timeout` would be interpreted as milliseconds instead of seconds.
+*   The incompatible flag
+    [`--incompatible_strict_action_env`](https://github.com/bazelbuild/bazel/issues/6648)
+    has been flipped to "false" again, as we discovered breakages for local
+    execution users. We'll need some more time to figure out the best way to
+    make this work for local and remote execution
+    ([#7026](https://github.com/bazelbuild/bazel/issues/7026)).
+*   The flag value `--test_sharding_strategy=experimental_heuristic` has been
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]: why removed? can we add a reason?

File: _posts/2019-02-26-bazel-0.23.md [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    removed.
+
+
+## Other changes
+
+*   Bazel binary is 20MB smaller by using a minimized embedded JDK.
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]: also list the % reduction?

File: _posts/2019-02-26-bazel-0.23.md [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+*   The embedded JDK version is upgraded to 11.0.2.
+*   Incompatible flag
+    [`--incompatible_auto_configure_host_platform`](https://github.com/bazelbuild/bazel/issues/7081)
+    has been added. When set, the default value of `--host_platform` and
+    `--platforms` will not be dependent on the configuration. This means that
+    setting `--cpu` or `--host_cpu` will not affect the target or host platform.
+*   Introduces
+    [`--local_ram_resources`](https://docs.bazel.build/versions/master/command-line-reference.html#flag--local_ram_resources)
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
    [`--local_ram_resources`](https://docs.bazel.build/versions/0.23.0/command-line-reference.html#flag--local_ram_resources)
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    and
+    [`--local_cpu_resources`](https://docs.bazel.build/versions/master/command-line-reference.html#flag--local_cpu_resources),
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
    [`--local_cpu_resources`](https://docs.bazel.build/versions/0.23.0/command-line-reference.html#flag--local_cpu_resources),
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    which will take the place of `--local_resources`.
+*   In
+    [`--keep_going`](https://docs.bazel.build/versions/master/user-manual.html#flag--keep_going)
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]:
```suggestion
    [`--keep_going`](https://docs.bazel.build/versions/0.23.0/user-manual.html#flag--keep_going)
```

File: _posts/2019-02-26-bazel-0.23.md [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+    repo using a pinned list of artifacts.
+*   Travis Cline created [`rules_homebrew`](https://github.com/tmc/rules_homebrew),
+    to manage brew packages.
+*   Wix engineer Natan Silnitsky wrote about their experience: [Migrating to
+    Bazel from Maven or Gradle? 5 crucial questions you should ask
+    yourself](https://medium.com/wix-engineering/migrating-to-bazel-from-maven-or-gradle-5-crucial-questions-you-should-ask-yourself-f23ac6bca070).
+*   Rahul Malik, iOS Platform Tech Lead, wrote about [developing fast & reliable
+    iOS builds at Pinterest (Part
+    one)](https://medium.com/@Pinterest_Engineering/developing-fast-reliable-ios-builds-at-pinterest-part-one-cb1810407b92).
+*   Christian Gruber wrote a blog post: [Bazely thinking and the tale of the
```
Comment (jin) [inline _posts/2019-02-26-bazel-0.23.md]: (not directly related to release notes) I think we should discuss with the larger team about the issues raised in this article. IIUC, it suggests that it's *okay* that Bazel picks an artifact from the repository cache based on the specified sha256, even if the URL does not match up.

File: _posts/2019-02-26-bazel-0.23.md (hunk 1 above)
Comment (laurentlb) [reply _posts/2019-02-26-bazel-0.23.md]: Is it useful to add this link?
https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/821

File: _posts/2019-02-26-bazel-0.23.md (hunk 11 above)
Comment (laurentlb) [reply _posts/2019-02-26-bazel-0.23.md]: I don't know

File: _posts/2019-02-26-bazel-0.23.md (hunk 1 above)
Comment (jin) [reply _posts/2019-02-26-bazel-0.23.md]: yes! do we have a fully green run?

Comment (laurentlb) [reply _posts/2019-02-26-bazel-0.23.md]: The tests for 0.23.0rc2 is fully green, 0.23.0rc3 has one failure.
Which one do you prefer?

File: _posts/2019-02-26-bazel-0.23.md (hunk 6 above)
Comment (laurentlb) [reply _posts/2019-02-26-bazel-0.23.md]: This one is available only at head.

File: _posts/2019-02-26-bazel-0.23.md (hunk 9 above)
Comment (laurentlb) [reply _posts/2019-02-26-bazel-0.23.md]: The version at head is better.

File: _posts/2019-02-26-bazel-0.23.md (hunk 1 above)
Comment (jin) [reply _posts/2019-02-26-bazel-0.23.md]: hmm, I think it doesn't matter that much. Let's go with the rc3.

**Comment by laurentlb** [comment]
I think it's now ready

**Pull request merged**